test/erc721Permit/ERC721Permit.permitForAll.t.sol

```diff
@@ -0,0 +1,352 @@
+
+        vm.startPrank(alice);
+        vm.expectRevert(UnorderedNonce.NonceAlreadyUsed.selector);
+        erc721Permit.permitForAll(alice, bob, true, block.timestamp, nonce, signature);
+        vm.stopPrank();
+    }
+
+    function test_fuzz_erc721permitForAll_unauthorized() public {
+        vm.prank(alice);
+        uint256 tokenId = erc721Permit.mint();
+
+        uint256 nonce = 1;
+        bytes32 digest = _getPermitForAllDigest(bob, true, nonce, block.timestamp);
+
+        // bob attempts signing an approval for himself
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(bobPK, digest);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        // approvals unset
+        assertEq(erc721Permit.isApprovedForAll(alice, bob), false);
+
+        // nonce was unspent
+        (uint256 wordPos, uint256 bitPos) = _getBitmapFromNonce(nonce);
+        assertEq(erc721Permit.nonces(alice, wordPos) & (1 << bitPos), 0);
+
+        vm.startPrank(bob);
+        vm.expectRevert(SignatureVerification.InvalidSigner.selector);
+        erc721Permit.permitForAll(alice, bob, true, block.timestamp, nonce, signature);
+        vm.stopPrank();
+
+        // approvals unset
+        assertEq(erc721Permit.isApprovedForAll(alice, bob), false);
+
+        // nonce was unspent
+        assertEq(erc721Permit.nonces(alice, wordPos) & (1 << bitPos), 0);
+    }
+
+    function test_fuzz_erc721permitForAll_deadlineExpired(address operator) public {
+        uint256 nonce = 1;
+        uint256 deadline = block.timestamp;
+        bytes32 digest = _getPermitForAllDigest(operator, true, nonce, deadline);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(alicePK, digest);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        // no approvals existed
+        assertEq(erc721Permit.isApprovedForAll(alice, operator), false);
+
+        // nonce was unspent
+        (uint256 wordPos, uint256 bitPos) = _getBitmapFromNonce(nonce);
+        assertEq(erc721Permit.nonces(alice, wordPos) & (1 << bitPos), 0);
+
+        // fast forward to exceed deadline
+        skip(1);
+
+        // -- PermitForAll but deadline expired -- //
+        vm.startPrank(operator);
+        vm.expectRevert(IERC721Permit_v4.DeadlineExpired.selector);
+        erc721Permit.permitForAll(alice, operator, true, deadline, nonce, signature);
+        vm.stopPrank();
+
+        // approvals unset
+        assertEq(erc721Permit.isApprovedForAll(alice, operator), false);
+
+        // nonce was unspent
+        assertEq(erc721Permit.nonces(alice, wordPos) & (1 << bitPos), 0);
+    }
+
+    /// @dev a signature for permit() cannot be used for permitForAll()
+    function test_fuzz_erc721Permit_invalidSignatureForAll(address operator) public {
+        vm.prank(alice);
+        uint256 tokenId = erc721Permit.mint();
+
+        uint256 nonce = 1;
+        uint256 deadline = block.timestamp;
+        bytes32 digest = keccak256(
```

Comment:
_getPermitDigest ?

Reply:
the helper function is only available in the sister test `ERC721Permit.permit.t.sol` :/